internal/state/state.go

```diff
@@ -1,50 +1,81 @@
 package state
 
 import (
-	"math/big"
-	"relayer/bindings/BridgeBase"
+	"relayer/internal/types"
+	"sort"
 
-	log "github.com/sirupsen/logrus"
-
-	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
 )
 
-// BlockchainClient defines the interface for blockchain operations required by getStateWithProof.
-type BlockchainClient interface {
-	LatestBlockNumber() (*big.Int, error)                                                  // Returns the block number
-	GetStateWithProof(opts *bind.CallOpts) (BridgeBase.SharedStructsStateWithProof, error) // Gets the state with proof
-	FinalizationInterval() *big.Int                                                        // Gets the finalization interval
+type State struct {
+	stakes      map[common.Address]int32
+	latestEpoch uint64
+	totalStake  int32
+	stakeGetter func(common.Address) int32
+}
+
+func NewState(getter func(common.Address) int32) *State {
+	return &State{
+		stakes:      make(map[common.Address]int32),
+		stakeGetter: getter,
+	}
 }
 
-// getStateWithProof fetches the state with proof for a given epoch and block number using the provided client.
-func GetStateWithProof(client BlockchainClient, logger *log.Logger, epoch *big.Int, block_num *big.Int) (*BridgeBase.SharedStructsStateWithProof, error) {
-	if block_num == nil {
-		block, err := client.LatestBlockNumber()
-		if err != nil || block == nil {
-			logger.WithField("block", block).WithError(err).Fatal("BlockByNumber")
+func (s *State) UpdateState(block *types.PillarBlock) {
+	if s.latestEpoch > uint64(block.Epoch) {
+		return
+	}
+	s.latestEpoch = uint64(block.Epoch)
+	for _, change := range block.VoteCountsChanges {
+		if _, ok := s.stakes[change.Address]; !ok {
+			stake := s.stakeGetter(change.Address)
+			s.stakes[change.Address] = stake
+			s.totalStake += stake
+		}
+		s.stakes[change.Address] += change.Value
+		s.totalStake += change.Value
+	}
+}
+
+type AccountWithSignature struct {
+	Address   *common.Address
+	Signature *types.CompactSignature
+}
+
+func (s *State) ReduceSignatures(block *types.PillarBlockData) ([]types.CompactSignature, error) {
+	var totalStake int32
+	accounts := make([]AccountWithSignature, 0, len(block.Signatures))
+	for _, signature := range block.Signatures {
```

Comment:
Possible to make it in one cycle if there will be sorted accounts slice in the state?